initial commit for embedding linker

adam-sutton-1992 opened the pull request:
This is WIP just showing the inital commit of the embedding linker, and how I've tested / instantiated it.

**WHAT IT DOES:**

This is a linker that requires no training. However it does have higher computational overheads.

Given an embedding model name (such as _sentence-transformers/all-MiniLM-L6-v2_ or _BAAI/bge-m3_) we embed all names in cdb.name2info. This will result in a N dimensional vector for each name. 

When provided with a bunch of entities from the ner step we then embed the detected names, or the raw texts and compare them to all embedded potential names via cosine similarity. 

With the most similar name we then obtain all cuis from that name via `cdb.name2info[predicted_name]["per_cui_status"].keys()`.

**HOW IT'S BEEN TESTED:**

I am currently changing the linker in an already existing model:

```from medcat.cat import CAT
from medcat.components.types import CoreComponentType
from medcat.components.linking.embedding_linker import Linker as EmbeddingLinker
from medcat.components.types import register_core_component

cat = CAT.load_model_pack(...)
register_core_component(CoreComponentType.linking, EmbeddingLinker.name, EmbeddingLinker.create_new_component)

embedding_linker = EmbeddingLinker(
    cdb=cat.cdb,
    config=cat.config,
    embedding_model_name="sentence-transformers/all-MiniLM-L6-v2",
    max_length=64
)
cat.config.components.linking.comp_name = EmbeddingLinker.name
cat._recreate_pipe()
embedding_linker = cat._pipeline._components[-1]
embedding_linker.embed_names(embedding_model_name="sentence-transformers/all-MiniLM-L6-v2", batch_size=4096)

cat.save_model_pack(...)
```

Just a note embedding with _sentence-transformers/all-MiniLM-L6-v2_ takes 10 minutes for 3 million names. _BAAI/bge-m3_ takes 1 hour 30. Although this is a one time cost.

I then run this to test performance:

```
from medcat.cat import CAT
from medcat.components.types import register_core_component
from medcat.components.types import CoreComponentType
from medcat.components.linking.embedding_linker import Linker as EmbeddingLinker
from medcat.stats.stats import get_stats
from medcat.data.mctexport import MedCATTrainerExport

register_core_component(CoreComponentType.linking, EmbeddingLinker.name, EmbeddingLinker.create_new_component)
cat = CAT.load_model_pack("/data/adam/models/kch_gstt_v2_embedding_linker")

**loading distemist and snomed_ct test sets**

all_projects = MedCATTrainerExport(projects=[])
all_projects["projects"].append(distemist)
all_projects["projects"].append(snomed_ct)
get_stats(cat2, all_projects)
```

I've tested this on the original model, and twice with the new embedding linker using "sentence-transformers/all-MiniLM-L6-v2" as the embedding model. Firstly, just chosing the first cui in self.cdb.name2info[predicted_name]["per_cui_status"].keys() if there are multiple. The second tries to do disambiguation with cui's prefered names and the detected name. 

Using context_based_linker:
Epoch: 0, Prec: 0.09023910064860668, Rec: 0.3306035738148675, F1: 0.14177920150470955

Picking first cui, with an embedding linker:

Epoch: 0, Prec: 0.07762498227887255, Rec: 0.3095756163585588, F1: 0.12412585010647798

With embedding disambiguation:

Epoch: 0, Prec: 0.0650762760285772, Rec: 0.2595346605900696, F1: 0.10406025987586687

The only real import stat here is recall. While these look low, I'm led to believe that these usually come with filters, which I've avoided. 

Issues / Discussion:

- I am currently storing the embeddings in cdb.name2info[name]["context_vectors"] for each name. This results in a 4.5gb larger model pack, and a loading time from 40 seconds -> 240 seconds. If possible it would be nice to be able to save a "context_matrix" where each row would correspond to the name (as long as name positioning persists). Maybe via the torch.save function although I'm not sure how to slot that in here.
- If the context based linker takes about 15 minutes on a project. The embedding linker takes 1hr30. Doing it entity by entity took about 6 hours. But doing the cosine similarity calculation for all detected names cut that time down. There might be more efficiencies to shave off here which I don't see... yet
- The performances are a little telling that I don't have disambiguation down yet. I'm open to suggestions. I am planning to try only embed according to cdb.cui2info[cui]['preferred_name']. So I don't have to use cdb.name2info[predicted_name]["per_cui_status"].keys(). I'm open to other suggestions maybe with name status itself.
- I'm not sure of a nice way to open up `embed_names`, besides addressing the component itself via protected variables:
```embedding_linker = cat._pipeline._components[-1]
- I don't save the models themselves locally to the model pack. We're not training the model pack, so just let them live in the huggingface hub repo with their buddies.

embedding_linker.embed_names(embedding_model_name="sentence-transformers/all-MiniLM-L6-v2", batch_size=4096)```

TODO:

1. Figure out the best linking / disambiguation method
2. Ensure lightweight saving/loading of models / vectors 
3. Create a config for this type of embedding. I think I only need to handle the embedding model, and the batch size used to generate embeddings. As they aren't exposed at the moment gracefully. 
4. Add whatever is required to make this a default component
5. Measure performance with other embedding methods (BAAI-M3)

Let me know any suggestions / whatever else I've missed. Thanks!

mart-r commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,467 @@
+        for names in tqdm(self._batch_data(cui_names, self.cnf_l.embedding_batch_size), total=total_batches, desc="Embedding cuis' preferred names"):
+            with torch.no_grad():
+                # removing ~ from names, as it is used to indicate a space in the CDB
+                names_to_embed = [name.replace("~", " ") for name in names]
```
The `prepare_name` method just iterates over tokens and ignore tokens that are marked to be skipped. The skipped words are ones that are in are marked as such by the [tagger](https://github.com/CogStack/cogstack-nlp/blob/main/medcat-v2/medcat/components/tagging/tagger.py) - that is they are either words in `config.preprocessing.words_to_skip` or punctuation, or stopwords (as defined by the tokenizer).
And the `prepare_name` method then just joins these tokens with the separator for subnames and names.

mart-r commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,467 @@
+        end_ind = entity.base.end_index
+
+        _left_tokens = doc[max(0, start_ind - size):start_ind]
+        tokens_left = [tkn for tkn in _left_tokens]
```
Yeah, that makes sense.

With the vector context model, we want to make sure we only include tokens for which we have a reasonable hope of having a context vector for. And punctuation and stop words won't have any context vectors in the vocab, so it makes sense to avoid that.

But because the model you're using is a more contextual one, it'll be able to make use of these details in a lot more general way - and will be able to derive context from the punctuation and stopwords as well.

adam-sutton-1992 commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,467 @@
+                if any(cui in include_cui_idxs for cui in name_cui_idxs)
+            ]
+            allowed_mask[torch.tensor(include_idxs, dtype=torch.long, device=self.device)] = True
+        else:
```
```
When no CUIs nor excluded CUIs are specified (the sets are empty),
    all CUIs are accepted.
    If there are CUIs specified then only those will be accepted.
    If there are excluded CUIs specified, they are excluded.
```

adam-sutton-1992 commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,467 @@
+
+        # score all detected contexts vs all names, handle in the loop each individual case
+        names_scores = detected_context_vectors @ self.names_context_matrix.T
+        cui_scores = detected_context_vectors @ self.cui_context_matrix.T
```
Fixed

adam-sutton-1992 commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,467 @@
+        similarity = candidate_scores[candidate_idx].item()
+        return predicted_cui, similarity
+
+    def _inference_by_names(
```
Fixed as per previous comments

adam-sutton-1992 commented:
TODO or discuss currently that I know of:

- Initialising for Inference I think having it embed upon init might not be the best way, as it can be time consuming. Especially if you're planning to change the embedding model. Altough handling is_dirty gracefuly is also required.
- Testing with removal of stop words / punctuation.
- Best way to embed names with seperators gracefully via reverse engineering the method or a suitable compromise.
- Optimal settings in the config as a base.
- Optimal similarity_threshold for precision / recall / f1.

Otherwise the changes I've submitted have done a few things:

- Added generating link candidates via names. If you choose to ignore candidates provided by the ner step, or have a detected entity with no candidates some are generated. This slightly improves performance in both cases, and is even better when combined with filtering.
- Removed infer by cui. With the generation of link candidates via names it feels quite obsolete. We only infer via cui names when disambiguating based on the longest name OR the preferred name (which can do with a bit more testing).
- Various bug fixes and improvements as discussed here.

mart-r commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,463 @@
+        for entity in all_ents:
+            if len(entity.link_candidates) == 1:
+                # if the include filter exists and the only cui is in it
+                if self.cnf_l.filters.cuis and entity.link_candidates[0] in self.cnf_l.filters.cuis:
```
There's a `LinkingFilters.check_filters` method for this. Should work for both this and the `elif`. You've used it above anyway.

In addition to that, there may be CUIs that have marked to always be disambiguated. These shouldn't be matched even if they're the only one. See https://github.com/CogStack/cogstack-nlp/blob/main/medcat-v2/medcat/components/linking/context_based_linker.py#L131
This can happen (e.g) if something has been annotated by the model and in supervised training the annotation has been marked as a false positive.

adam-sutton-1992 replied:
Oops, fixed.

mart-r reviewed (commented):
Some small comments on the code

mart-r commented:
> Initialising for Inference I think having it embed upon init might not be the best way, as it can be time consuming. Especially if you're planning to change the embedding model. Altough handling is_dirty gracefuly is also required.

You could also init at `__call__` time when it's first needed. But that's really quite unintuitive since it would mean the first inference call would take "forever" - even if it was for one small piece of text. Which could cause the user to think it's broken.
Like I said before though - as long as you provide a tutorial alongside that shows what the initial step(s) are to get this saved on disk, it'll probably be fine. So if the initial model creation involves an additional step (after setting config and stuff) to do the embedding, and save the model after that, I'd say that would be good enoguh.

As for `is_dirty`, I think the best you can do is to check for the flag and log a warning. Because you could recalculate at `__call__` time if you find the CDB to be dirty, but this may - again - take so long that the user thinks the whole thing is broken. Though if the change is addative (in terms of CUIs or or names) then you should be able to only embed the bits that were added. But things may not be as easy if there's stuff removed (though probably still technically doable).
The other option would be to overwrite some of the CDB's methods (to add / remove cuis / names) and either log a warning at that time or even raise an exception (this could also be a configuration option).

> Best way to embed names with seperators gracefully via reverse engineering the method or a suitable compromise.

I don't think there's really a way to do that gracefully. This would be a generative process - you're wanting to generate information that isn't there.
The only real way I could see this working is if you had the SNOMED release (and the UMLS bits that were used for enrichment) and then just mapped all the processed names to their originals. You could do this as a one time thing on a per model basis, but it'd be tedious to do in a general manner.



Overall, you migth also want to add some tests to this. Probably with a very low-performant but small model. But something that shows the entire thing works in the technical sense.

adam-sutton-1992 commented:
I've just this in the __call__ method:

if self.cdb.is_dirty:
            logging.warning("CDB has been modified since last save/load. This might significantly affect linking performance.")
            logging.warning("If you have added new concepts or changes, please re-embed the CDB names and cuis before linking.")

I think there are tons are permutations I can do around being smart with embedding names and cuis. But I think it might just not be worth it.

mart-r commented:
I think that's fair.

Though maybe also tell them the method to use for the re-embedding. I.e:
```python
linker = self.cat._pipeline.get_component(CoreComponentType.linking)
linker.embed_cui_names(linker.embedding_model_name)
```
PS: I'd probably need to improve getting the pipe from the CAT object instead of using the protected attribute.

mart-r commented:
The API to access the component would be:
```python
from medcat.components.types import CoreComponentType
comp = cat.pipe.get_component(CoreComponentType.linking)
```

adam-sutton-1992 commented:
Yep. That works nicely.

mart-r commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,583 @@
+            logger.debug("Using the same model for embedding names.")
+        else:
+            self.cnf_l.embedding_model_name = embedding_model_name
+        names = list(self.cdb.name2info.keys())
```
Why not reuse `self._name_keys`?

mart-r commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,583 @@
+            device=self.device,
+        )
+        self._valid_names = _has_cuis_all & allowed_mask
+        self._last_include_set = include_set
```
You probably want a copy of these. Otherwise changes to the set (i.e adding, removing elements) wouldn't result in a recalculation here because it's the same instance as the "last" ones.
Currently, if someone wrote a fully new filter (either exclude or include), the recalculation would be triggered, but not if someone adds / removes stuff to / from existing sets.

mart-r commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,583 @@
+        sorted_indices = torch.argsort(names_scores, dim=1, descending=True)
+
+        for i, entity in enumerate(entities):
+            link_candidates = [
```
There's a config option `config.components.linking.filter_before_disamb` that should (generally) be checked before filtering candidates before disambiguation.

You generally want to filter for the CUIs _after_ you've found the suitable CUI. That's because otherwise (in case of a filter) you end up never considering other CUIs (that may be a closer match) for the same name. For instance, `fit` could refer to `91175000` (`Seizure`) or `102499006` (`Fit and well`). However, if you filter out one of them before disambiguation, you'll be guaranteed to get the other as the "best fit". And unless the threshold is quite aggressive, this may very well be considered a valid output.

PS:
I wouldn't be surprised if I asked you to do this at some point in the past in order to do less work. So if I did, sorry for that.

adam-sutton-1992 replied:
Resolved

mart-r commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,583 @@
+                cuis = list(self.cdb.name2info[detected_name]["per_cui_status"].keys())
+
+                predicted_cui, _ = self._disambiguate_by_cui(cuis, cui_scores[i, :])
+
```
There should be a check for the filters here - if the "best match" isn't allowed, don't use it.

See comment from above for more detail.

mart-r commented on medcat-v2/medcat/components/linking/embedding_linker.py:
```diff
@@ -0,0 +1,583 @@
+                    entity.context_similarity = 1
+                    le.append(entity)
+                    continue
+            # it has to be inferred due to filters or number of link candidates
```
I feel like if there's only 1 link candidate (and it's filtered out) then there's generally no reason to send the entity for inference. That only really makes sense if you don't trust the NER process at all. At the very least, I'd check (in an `elif` clause above within the 1-candidate clause) for `self.cnf_l.use_ner_link_candidates` and if that's true, `continue` (i.e avoid sending this to inference).

adam-sutton-1992 replied:
Done - I think!

adam-sutton-1992 commented on medcat-v2/medcat/config/config.py:
```diff
@@ -376,6 +376,41 @@ class Linking(ComponentConfig):
     class Config:
         extra = 'allow'
 
+class EmbeddingLinking(Linking):
```
Done

adam-sutton-1992 commented on medcat-v2/tests/components/linking/test_embedding_linker.py:
```diff
@@ -0,0 +1,66 @@
+
+class EmbeddingLinkerInitTests(ComponentInitTests, unittest.TestCase):
+    expected_def_components = 4
+    comp_type = types.CoreComponentType.linking
```
done

adam-sutton-1992 commented on medcat-v2/tests/components/linking/test_embedding_linker.py:
```diff
@@ -0,0 +1,66 @@
+        cls.vtokenizer = FakeTokenizer()
+        cls.comp_cnf = getattr(cls.cnf.components, cls.comp_type.name)
+
+    def test_can_create_def_component(self):
```
done

adam-sutton-1992 commented on medcat-v2/tests/components/linking/test_embedding_linker.py:
```diff
@@ -0,0 +1,66 @@
+        registered_names = [name for name, _ in avail_components]
+        self.assertIn("medcat2_embedding_linker", registered_names)
+
+class TrainableEmbeddingLinkerTests(unittest.TestCase):
```
Done, I've had to give additional attributes to the fake cdb and document. but I can pass through with no entities. Unless you want me to test with actual entities?

adam-sutton-1992 commented on medcat-v2/tests/components/linking/test_embedding_linker.py:
```diff
@@ -0,0 +1,66 @@
+
+
+class EmbeddingLinkerInitTests(ComponentInitTests, unittest.TestCase):
+    expected_def_components = 4
```
Done

mart-r commented on medcat-v2/medcat/config/config.py:
```diff
@@ -377,6 +377,10 @@ class Config:
         extra = 'allow'
 
 class EmbeddingLinking(Linking):
+    """Changing compoenent name"""
```
I got it working again locally (I think my VSCode's language server had died). Here's what it looks like for me for the other configs:
<img width="604" height="72" alt="image" src="https://github.com/user-attachments/assets/10b8a17e-c695-4398-be5a-cd7f4b56d9d6" />

So if you have the comments before, you'd see the wrong one (i.e the one for the one after).
